packages/world/src/systems/BedSystem.sol

```diff
@@ -93,38 +81,50 @@ contract BedSystem is System {
   function removeDeadPlayerFromBed(EntityId player, Vec3 dropCoord) public {
     checkWorldStatus();
 
-    EntityId bed = PlayerBed._getBedEntityId(player);
-    require(bed._exists(), "Player is not in a bed");
-
-    Vec3 bedCoord = bed._getPosition();
+    (EntityId bed, Vec3 bedCoord) = BedLib.getPlayerBed(player);
 
     require(bedCoord.inSurroundingCube(dropCoord, MAX_RESPAWN_HALF_WIDTH), "Drop location is too far from bed");
 
     (EntityId drop, ObjectType objectType) = EntityUtils.getOrCreateBlockAt(dropCoord);
     require(objectType.isPassThrough(), "Cannot drop items on a non-passable block");
 
-    (EntityId forceField, EntityId fragment) = ForceFieldUtils.getForceField(bedCoord);
-
-    uint128 depletedTime = decreaseFragmentDrainRate(forceField, fragment, PLAYER_ENERGY_DRAIN_RATE);
-    EnergyData memory playerData = BedLib.updateSleepingPlayer(player, bed, depletedTime, bedCoord);
+    EnergyData memory playerData = BedLib.removePlayerFromBed(player, bed, bedCoord);
     require(playerData.energy == 0, "Player is not dead");
 
     BedLib.transferInventory(player, drop);
-
-    PlayerUtils.removePlayerFromBed(player, bed);
   }
 }
 
 // To avoid reaching bytecode size limit
 library BedLib {
-  function transferInventory(EntityId player, EntityId bed) public {
-    InventoryUtils.transferAll(player, bed);
+  function addPlayerToBed(EntityId caller, EntityId bed, Vec3 bedCoord) public {
+    (EntityId forceField, EntityId fragment) = ForceFieldUtils.getForceField(bedCoord);
+    require(forceField._exists(), "Bed is not inside a forcefield");
+
+    uint128 depletedTime = increaseFragmentDrainRate(forceField, fragment, PLAYER_ENERGY_DRAIN_RATE);
+    PlayerBed._setBedEntityId(caller, bed);
+    BedPlayer._set(bed, caller, depletedTime);
+  }
+
+  function removePlayerFromBed(EntityId caller, EntityId bed, Vec3 bedCoord) public returns (EnergyData memory) {
+    (EntityId forceField, EntityId fragment) = ForceFieldUtils.getForceField(bedCoord);
+    uint128 depletedTime = decreaseFragmentDrainRate(forceField, fragment, PLAYER_ENERGY_DRAIN_RATE);
+
+    EnergyData memory playerData = updateSleepingPlayerEnergy(caller, bed, depletedTime, bedCoord);
+
+    PlayerUtils.removePlayerFromBed(caller, bed);
+    return playerData;
   }
 
-  function updateSleepingPlayer(EntityId player, EntityId bed, uint128 depletedTime, Vec3 bedCoord)
-    public
-    returns (EnergyData memory)
-  {
-    return updateSleepingPlayerEnergy(player, bed, depletedTime, bedCoord);
+  function getPlayerBed(EntityId player) public view returns (EntityId, Vec3) {
+    EntityId bed = PlayerBed._getBedEntityId(player);
+    require(bed._exists(), "Player is not in a bed");
+
+    Vec3 bedCoord = bed._getPosition();
+    return (bed, bedCoord);
+  }
+
+  function transferInventory(EntityId player, EntityId bed) public {
+    InventoryUtils.transferAll(player, bed);
```

Comment:
this confused me for a bit cuz I thought we were transferring to air, and we are

Could we just remove this util in bedlib and call `InventoryUtils.transferAll(player, drop);` directly?

Reply:
bed system is almost over the gas limit :/ but I'll rename the args

Reply:
I think `transferAll` is already public though, so it shouldnt change gas no?